src/API/Reports/Categories/DataStore.php

```diff
@@ -133,7 +133,7 @@ protected function add_order_by_params( $query_args, $from_arg, $id_cell ) {
 		if ( false !== strpos( $order_by_clause, '_terms' ) ) {
 			$join = "JOIN {$wpdb->terms} AS _terms ON {$id_cell} = _terms.term_id";
 			if ( 'inner' === $from_arg ) {
-				$this->subquery->add_sql_clause( 'join', $join );
+				$this->subquery->add_sql_clause( 'left_join', $join );
```

Comment:
This is a deviation from the pre-refactor SQL - does this have the potential to change the result set? In my research of this issue, the difference I saw before and after the refactor was the order of the join statements, not the type of join. Pre-refactor the `LEFT JOIN`s occurred before the (inner) `JOIN`s - whereas on `master`, the order is reversed.

Is that ordering change not the regression/bug?

Additionally, this would be a *great* case to cover with a new test.

Reply:
> not the type of join

The array key has no effect on the type of join in the query. It affects the order that the query is concatenated in. The original queries were concatenated using 3 array keys. The refactor kept the keys but gave them better names to reflect the type of join one would expect to find in that element. This query has an exception where an [inner] join occurs after left joins.

With reformatting, this is a query from `eb1c5a3cc1150d15ee78a9b7758bd8077b7c71c5` before the refactor was merged

```
SELECT
	wp_wc_category_lookup.category_tree_id as category_id,
	SUM(product_qty) as items_sold, 
	SUM(product_net_revenue) AS net_revenue, 
	COUNT(DISTINCT wp_wc_order_product_lookup.order_id) as orders_count, 
	COUNT(DISTINCT wp_wc_order_product_lookup.product_id) as products_count
FROM
	wp_wc_order_product_lookup
	 LEFT JOIN wp_term_relationships ON wp_wc_order_product_lookup.product_id = wp_term_relationships.object_id 
	 LEFT JOIN wp_wc_category_lookup ON wp_term_relationships.term_taxonomy_id = wp_wc_category_lookup.category_id 
	 JOIN wp_terms AS _terms ON wp_wc_category_lookup.category_tree_id = _terms.term_id 
	 JOIN wp_wc_order_stats ON wp_wc_order_product_lookup.order_id = wp_wc_order_stats.order_id
WHERE
	1=1
	 AND wp_wc_order_product_lookup.date_created <= '2019-12-31 23:59:59' 
	 AND wp_wc_order_product_lookup.date_created >= '2019-10-01 00:00:00'
	 AND ( wp_wc_order_stats.status NOT IN ( 'wc-trash','wc-pending','wc-cancelled','wc-failed' ) ) 
	 AND wp_wc_category_lookup.category_tree_id IS NOT NULL
GROUP BY
	wp_wc_category_lookup.category_tree_id
ORDER BY
	_terms.name desc
```

The same query from this branch

```
SELECT
	wp_wc_category_lookup.category_tree_id as category_id, 
	SUM(product_qty) as items_sold, 
	SUM(product_net_revenue) AS net_revenue, 
	COUNT(DISTINCT wp_wc_order_product_lookup.order_id) as orders_count, 
	COUNT(DISTINCT wp_wc_order_product_lookup.product_id) as products_count
FROM
	wp_wc_order_product_lookup
	JOIN wp_wc_order_stats ON wp_wc_order_product_lookup.order_id = wp_wc_order_stats.order_id 
	LEFT JOIN wp_term_relationships ON wp_wc_order_product_lookup.product_id = wp_term_relationships.object_id 
	LEFT JOIN wp_wc_category_lookup ON wp_term_relationships.term_taxonomy_id = wp_wc_category_lookup.category_id 
	JOIN wp_terms AS _terms ON wp_wc_category_lookup.category_tree_id = _terms.term_id
WHERE
	1=1
	AND ( wp_wc_order_stats.status NOT IN ( 'wc-trash','wc-pending','wc-cancelled','wc-failed' ) ) 
	AND wp_wc_category_lookup.category_tree_id IS NOT NULL 
	AND wp_wc_order_product_lookup.date_created <= '2019-12-31 23:59:59' 
	AND wp_wc_order_product_lookup.date_created >= '2019-10-01 00:00:00'
GROUP BY
	wp_wc_category_lookup.category_tree_id
ORDER BY
	_terms.name desc
```

> Additionally, this would be a great case to cover with a new test.

Sounds good.

Reply:
> The array key has no effect on the type of join in the query. It affects the order that the query is concatenated in. 

Ah, right - I missed that. Since we're hacking the join "type" (as far as the query class knows) for the purposes of statement ordering, a comment documenting the reasoning would be extremely helpful.

> This query has an exception where an [inner] join occurs after left joins.

Is it the outlier in all of our reports? My original plan of attack for this issue was to verify that the "right, inner, left" order was correct/consistent with the pre-refactor queries.

> The same query from this branch

Looks like we've got one inner join still occurring in a different order.

Reply:
> Looks like we've got one inner join still occurring in a different order.

That one is okay because it joins to the main table in the query and isn't used by any of the other joins.

Reply:
> Additionally, this would be a great case to cover with a new test.
> a comment documenting the reasoning would be extremely helpful.

Added.

> My original plan of attack for this issue was to verify that the "right, inner, left" order was correct/consistent with the pre-refactor queries.

The MySQL query optimizer will restructure/reorganize queries so I would expect that if the clauses are the same and the query executes then the order that they appear would not affect the results. `OR` `WHERE` conditions would be an exception if they were not wrapped in parentheses.

Should we add unit tests so the data stores are covered for sorting, filtering, etc. ?

Reply:
> The MySQL query optimizer will restructure/reorganize queries so I would expect that if the clauses are the same and the query executes then the order that they appear would not affect the results. OR WHERE conditions would be an exception if they were not wrapped in parentheses.

Ok, seems like that would be provable with an `EXPLAIN` then?

> Should we add unit tests so the data stores are covered for sorting, filtering, etc. ?

Yes please! Totally fine for follow up PR(s).

Reply:
> seems like that would be provable with an EXPLAIN then?

Here's the `EXPLAIN` for both

### original

```
id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
1	SIMPLE	wp_wc_order_product_lookup	NULL	range	order_id,product_id,date_created	date_created	5	NULL	324	100.00	Using index condition; Using temporary; Using filesort
1	SIMPLE	wp_wc_order_stats	NULL	eq_ref	PRIMARY,status	PRIMARY	8	ron.wp_wc_order_product_lookup.order_id	1	60.00	Using where
1	SIMPLE	wp_term_relationships	NULL	ref	PRIMARY,term_taxonomy_id	PRIMARY	8	ron.wp_wc_order_product_lookup.product_id	2	100.00	Using index
1	SIMPLE	wp_wc_category_lookup	NULL	index	PRIMARY	PRIMARY	16	NULL	47	9.00	Using where; Using index; Using join buffer (Block Nested Loop)
1	SIMPLE	_terms	NULL	eq_ref	PRIMARY	PRIMARY	8	ron.wp_wc_category_lookup.category_tree_id	1	100.00	NULL
```

### refactored

```
id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
1	SIMPLE	wp_wc_order_product_lookup	NULL	range	order_id,product_id,date_created	date_created	5	NULL	324	100.00	Using index condition; Using temporary; Using filesort
1	SIMPLE	wp_wc_order_stats	NULL	eq_ref	PRIMARY,status	PRIMARY	8	ron.wp_wc_order_product_lookup.order_id	1	60.00	Using where
1	SIMPLE	wp_term_relationships	NULL	ref	PRIMARY,term_taxonomy_id	PRIMARY	8	ron.wp_wc_order_product_lookup.product_id	2	100.00	Using index
1	SIMPLE	wp_wc_category_lookup	NULL	index	PRIMARY	PRIMARY	16	NULL	47	9.00	Using where; Using index; Using join buffer (Block Nested Loop)
1	SIMPLE	_terms	NULL	eq_ref	PRIMARY	PRIMARY	8	ron.wp_wc_category_lookup.category_tree_id	1	100.00	NULL
```